clarity-ext-scripts/clarity_ext_scripts/covid/partner_api_client.py

```diff
@@ -255,12 +266,101 @@ def search_for_service_request(self, org, org_referral_code):
+
+    def _create_payload(self, service_request_id, diagnosis_result, analysis_results):
+        # TODO Need to think about if this needs refactoring later, to more easily support multiple
+        #      analysis types. This is going to be rather unwieldy to add more as it is implemented
+        #      atm.
+        observations = self._create_observations(analysis_results)
+        diagnosis_result_as_codeable_concept = self._translate_diagnosis_result_to_codeable_concept(
+            diagnosis_result)
+        return self._create_diagnosis_report_object(service_request_id=service_request_id,
+                                                    observations=observations,
+                                                    codeable_concept=diagnosis_result_as_codeable_concept)
+
+    def _create_observations(self, analysis_results):
+        # TODO Need to decide on format for the analysis results to be gathered in.
+        observations = []
+        for index, result in enumerate(analysis_results):
+            observations.append(
+                self._create_observation(index, result["value"]))
+        return observations
+
+    def _create_observation(self, index, value):
+        # TODO Note this this now hard codes the observation type. We will need to
+        #      extend this if we implement other types of analysis.
+        return {
+            "resourceType": "Observation",
+            "id": str(index),
+            "status": "final",
+            "code": "v1-ct-value-mgi-real-time-fluorescent-RT-PCR-2019-nCoV",
+            "system": "http://uri.ctmr.scilifelab.se/id/CodeSystem/cs-observations",
+            "valueQuantity": {
+                "value": value
+            }
+        }
+
+    def _translate_diagnosis_result_to_codeable_concept(self, diagnosis_result):
+        if diagnosis_result not in VALID_COVID_RESPONSES:
+            raise AssertionError("Diagnosis result {} no in list of valid resonses: {}.".format(
+                diagnosis_result, VALID_COVID_RESPONSES
+            ))
+        return {
+            "coding": [
+                {
+                    "system": "{}/id/CodeSystem/cs-result-types".format(self._test_partner_code_system_base_url),
+                    "code": diagnosis_result
+                }
+            ]
+        }
+
+    def _create_diagnosis_report_object(self, service_request_id, observations, codeable_concept):
+        # TODO Do we not need to use the referral code? Is it really just the service_request_id
+        #      we need to use? I can't see where it is supposed to go.
+        return {
+            "resourceType": "DiagnosticReport",
+            "contained": observations,
+            "basedOn": [
+                {
+                    "reference": "ServiceRequest/{}".format(service_request_id)
+                }
+            ],
+            "code": codeable_concept,
+            "result": map(lambda x: {"reference": "Observation/#{}".format(x["id"])}, observations)
```

Comment:
Or as list comprehension? I've grown to thinking it's more readable :) but that's a matter of taste.

Reply:
I would prefer that the variable x would just be called `observation`, although you can probably find a billion examples of me using a one letter variable in list comprehensions.

Reply:
I'm gonna stick to my functional programming tendencies on this one. :wink: I don't mind anyone else changing it later, but since I know this works, I'd prefer to keep it this way for now.

Reply:
Sounds good!

(But I think list comprehensions are just as much functional programming, just with more beautiful syntax (personal preference).

Sorry for mentioning this kind of thing given the circumstances ;) )

Reply:
... i.e. they hide the ugly `lambda` keyword.